theano/scan_module/scan_opt.py

```diff
@@ -393,143 +423,159 @@ def apply(self, fgraph):
             self.process_node(fgraph, node)
 
     def process_node(self, fgraph, node):
+        """
+        IMPORTANT NOTE: This function uses set and dictionary data structure.
+        By default they are not ordered for efficiency reasons. Take care
+        and make sure of changing them to Ordered versions if you need to
+        iterate over those variables.
+        """
         # this flag tells if there was any change during the last iterations
-        changed = True
         clean_inputs, clean_outputs = scan_utils.reconstruct_graph(
             node.op.inputs, node.op.outputs)
 
-        local_fgraph = gof.FunctionGraph(clean_inputs, clean_outputs, clone=False)
-        max_iterations = 2 * len(local_fgraph.toposort()) + 3
-        counts = 0
-        to_remove = []
-        to_replace = []
+        local_fgraph = gof.FunctionGraph(clean_inputs, clean_outputs,
+                                         clone=False)
+        local_fgraph_topo = local_fgraph.toposort()
+        local_fgraph_outs_set = set(local_fgraph.outputs)
+        local_fgraph_outs_map = dict([(v,k) for k,v in \
+                                     enumerate(local_fgraph.outputs)])
+
+        to_remove_set = set()
+        to_remove_add = to_remove_set.add
+        to_replace_set = set()
+        to_replace_add = to_replace_set.add
+        to_replace_map = OrderedDict()
+        nto_replace = 0
+
+        def add_to_replace(y, nto_replace):
+            to_replace_add(y)
+            to_replace_map[y] = nto_replace
+            return nto_replace + 1
```

Comment:
This function should probably not take nto_replace as an input and return it as an output; it not needed and if someone forgets to use the output, he's in big trouble. The fonction can instead use nto_replace directly and modify it inplace. This will make the whole thing a lot easier to use.